toolkit/tools/graphPreprocessor/graphPreprocessor.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"os"
+
+	"gopkg.in/alecthomas/kingpin.v2"
+	"microsoft.com/pkggen/internal/exe"
+	"microsoft.com/pkggen/internal/logger"
+	"microsoft.com/pkggen/internal/pkggraph"
+)
+
+const (
+	allGoalNodeName = "ALL"
+)
+
+var (
+	app             = kingpin.New("graphPreprocessor", "Update the graph for the build requested")
+	inputGraphFile  = exe.InputFlag(app, "Input graph file having full build graph")
+	outputGraphFile = exe.OutputFlag(app, "Output file to export the scrubbed graph to")
+	imageConfig     = app.Flag("image-config-file", "Optional image config file to extract a package list from.").String()
+	baseDirPath     = app.Flag("base-dir", "Base directory for relative file paths from the config. Defaults to config's directory.").ExistingDir()
+	hydratedBuild   = app.Flag("hydrated-build", "Build individual packages with dependencies Hydrated").Bool()
+	ignoredPackages = app.Flag("ignored-packages", "Space separated list of specs ignoring rebuilds if their dependencies have been updated. Will still build if all of the spec's RPMs have not been built.").String()
+	pkgsToBuild     = app.Flag("packages", "Space separated list of top-level packages that should be built. Omit this argument to build all packages.").String()
+	pkgsToRebuild   = app.Flag("rebuild-packages", "Space separated list of base package names packages that should be rebuilt.").String()
+
+	logFile  = exe.LogFileFlag(app)
+	logLevel = exe.LogLevelFlag(app)
+)
+
+func replaceRunNodesWithPrebuiltNodes(pkgGraph *pkggraph.PkgGraph) (err error) {
+	for _, node := range pkgGraph.AllNodes() {
+
+		if node.Type != pkggraph.TypeRun {
+			continue
+		}
+
+		isPrebuilt, _ := pkggraph.IsSRPMPrebuilt(node.SrpmPath, pkgGraph, nil)
+
+		if isPrebuilt == false {
+			continue
+		}
+
+		preBuiltNode := pkgGraph.CloneNode(node)
+		preBuiltNode.State = pkggraph.StateUpToDate
+		preBuiltNode.Type = pkggraph.TypePreBuilt
+
+		parentNodes := pkgGraph.To(node.ID())
+		for parentNodes.Next() {
+			parentNode := parentNodes.Node().(*pkggraph.PkgNode)
+
+			if parentNode.Type != pkggraph.TypeGoal {
+				pkgGraph.RemoveEdge(parentNode.ID(), node.ID())
+
+				logger.Log.Debugf("Adding a 'PreBuilt' node '%s' with id %d. For '%s'", preBuiltNode.FriendlyName(), preBuiltNode.ID(), parentNode.FriendlyName())
+				err = pkgGraph.AddEdge(parentNode, preBuiltNode)
+
+				if err != nil {
+					logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
```

Comment:
We should return in case of an error.

Reply:
Btw. it looks like I've missed doing that as well in `pkggraph.go::fixPrebuiltSRPMsCycle()`.

Reply:
```suggestion
					logger.Log.Errorf("Adding edge failed for %v -> %v", parentNode, preBuiltNode)
					return
```